### ISSUE TITLE
How to ensure the model being trained on all the users and items?

### ISSUE THREAD

**Issue by jpzhangvincent** [massquantity/LibRecommender#423]
Is there a split function to make sure the model being trained on all the users and items? Instead of random split, we want to make the all the embedings learned and available for all the users and items during training, while we can still evaluate on the validation/test set.

**Comment by massquantity** [comment]
After splitting, you can manually move all the oov(out-of-vocabulary) users/items from eval_data to train_data.

```python
train_data, eval_data = split_by_ratio(...)
train_data, eval_data = move_oov_to_train(train_data, eval_data)

def move_oov_to_train(train_data: pd.DataFrame, eval_data: pd.DataFrame):
    unique_users = np.unique(train_data["user"])
    unique_items = np.unique(train_data["item"])
    eval_user_oov = np.isin(eval_data["user"].to_numpy(), unique_users, invert=True)
    eval_item_oov = np.isin(eval_data["item"].to_numpy(), unique_items, invert=True)
    oov_mask = eval_user_oov | eval_item_oov
    train_data = pd.concat([train_data, eval_data[oov_mask]], axis=0)
    eval_data = eval_data[~oov_mask]
    return train_data, eval_data
```

It is better to retrain using all the combined data after initial training.

**Comment by massquantity** [comment]
OK, I'll consider adding this extension.

"retrain" in this library means incremental training, which is used when one gets some new data potentially several days later. So the previous trained model has to be saved.

In your case, the "retrain" feature is not needed. What i meant in the previous reply is using all the data(train + validation + test) and training again with the same hyper-parameters after the initial training.

**Comment by jpzhangvincent** [comment]
Another question, since for the ranking task with implicit data which only contains positive examples, how can we still get the learned embedding for those items that are not being included in the data (because they are not being transacted)?

**Comment by massquantity** [comment]
The embeddings of users/items with no transaction can't be learned, since they have no data to be trained. The library applies mean embedding, so all these users will have the same embedding.

```python
user_embeds = model.get_user_embedding()  # shape: (n_users, embed_size)
no_transaction_user_embed = np.mean(user_embeds, axis=0)

**Comment by jpzhangvincent** [comment]
Hmm I thought the model can still infer the embeddings for the new or long-tail items if we pass the features that are generalizable across the item/user set.. since the two tower model can learn the representation/interaction of the features as a content based approach.. Correct me if I’m wrong? Are you saying we can’t use the explicit data(with both positive and negative samples) for the ranking task?

**Comment by massquantity** [comment]
OK, i get it. You didn't mention that you can use other features in your data. This is supported by using [Dynamic Embedding Generation](https://librecommender.readthedocs.io/en/latest/user_guide/embedding.html#dynamic-embedding-generation).
However, this approach only supports new users. *Maybe* the item embedding generation will be added in the future.
